### PR TITLE
[0.63] Debugging: Adding variables to watch window crashes the process hosting the addin VSO 4853697

### PR DESCRIPTION
**Pull request by mganandraj**
The crash is happening when the native logging callback gets called during evaluating scripts from debugger.

Native callbacks in ChakraExecutor are expected to be called only when running scripts initiated from the executor by evaluating a script or calling a function, where we set the JSContextInfo in stack with the active ChakraCore context. This allows us to reuse the same thread/runtime across multiple SDX instances/contexts. I'm not sure anyone is really using the thread/runtime sharing, hence we could fix it by making the JSContextInfo a thread local variable. But, it's a much riskier fix hence avoiding it.

The crash should've been happening from the start when executing 'console.log("abcd")' from watch window. But, that's not a usual user scenario hence went unnoticed very likely. Recently, something changed somewhere (sorry, I couldn't yet figure out), so that the debugger is replaying the past console log calls whenever a script is evaluated in the watch windows. This is happening even when console redirection is turned off.

This change is fixing the crash by ensuring that an instance of JSContextHolder is active in the stack with the executor instance, before dispatching the callback to the implementation.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7449)

### CODE REVIEW

**Review by NickGerleman — APPROVED** [review]

File: vnext/Chakra/ChakraExecutor.cpp [hunk 1]
```diff
@@ -209,20 +213,20 @@ JsErrorCode ChakraExecutor::RedirectConsoleToDebugger(JsValueRef debuggerConsole
 
   const char *script =
```
Comment (asklar) [inline vnext/Chakra/ChakraExecutor.cpp:214]: this can be constexpr and make sure it goes into .rdata

File: vnext/Chakra/ChakraExecutor.cpp [hunk 2]
```diff
@@ -209,20 +213,20 @@ JsErrorCode ChakraExecutor::RedirectConsoleToDebugger(JsValueRef debuggerConsole
...
+              "obj[name] = function(...args) {\n"
+              "console[name](...args);\n"
+              "if (name in debugConsole && typeof debugConsole[name] === \"function\") {\n"
+                "debugConsole[name](...args);\n"
+              "}\n"
+            "}\n"
+            "})(fn);\n"
+          "}\n"
+        "}\n"
+        "global.console = obj;\n"
```
Comment (asklar) [inline vnext/Chakra/ChakraExecutor.cpp:229]: nit, could get rid of the \n and spaces since this isn't designed to be used/seen by humans

**Review by asklar — APPROVED** [review]

File: vnext/Chakra/ChakraExecutor.cpp (hunk 2 above)
Comment (asklar) [reply vnext/Chakra/ChakraExecutor.cpp:229]: also you'll need to run yarn format which will remove the leading spaces before the string starts too

**Comment by asklar** [comment]
please run `yarn change --branch upstrea/master` to create a change file

**Comment by asklar** [comment]
Please make sure that this gets merged into master and 64 too (ideally first)

**Comment by NickGerleman** [comment]
> Please make sure that this gets merged into master and 64 too (ideally first)

I don't think anything in 0.64 should be using ChakraExecutor, so this is _probably_ fine staying here.

**Comment by mganandraj** [comment]
> Please make sure that this gets merged into master and 64 too (ideally first)

As Nick mentioned below, ChakraExecutor shouldn't be a thing in RN64. [PR](https://github.com/microsoft/react-native-windows/commit/1c6b617fcecefd9b873da00af594c71e61d4e792) has already removed the source from RN64.

**Pull request merged**